File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
# 백준 12865 평범한 배낭

N, K = map(int, input().split())

dp = [[0 for _ in range(K+1)] for _ in range(N+1)]

bag = [[0,0]]
for _ in range(N):
    bag.append(list(map(int,input().split())))


for i in range(1,N+1): # 물건
    for j in range(1,K+1): # 가방 검사
        weight = bag[i][0]
        value = bag[i][1]
        
        if j < weight:
            dp[i][j] = dp[i-1][j]
        else:
            dp[i][j] = max(value + dp[i-1][j-weight],dp[i-1][j])
        
print(dp[N][K])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
// 백준 12865번 평법한 배낭

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.stream.Stream;

class Main{
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int[] nums = Stream.of(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int N = nums[0];
        int K = nums[1];
        int[][] items = new int[N][2];
        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());;
            items[i][0] = Integer.parseInt(st.nextToken());
            items[i][1] = Integer.parseInt(st.nextToken());
        }

        int[][] dp = new int[N+1][K+1];

        for (int i = 1; i < N+1; i++) {
            for (int j = 1; j < K+1; j++) {
                int weight = items[i-1][0];
                int value = items[i-1][1];

                if(j < weight){
                    dp[i][j] = dp[i-1][j];
                }else{
                    dp[i][j] = Math.max(dp[i-1][j], dp[i-1][j-weight] + value);
                }
            }
        }

        System.out.println(dp[N][K]);

    }
}